co-simulation tool/fed/mosaic-carla/src/main/java/org/eclipse/mosaic/fed/carla/ambassador/CarlaAmbassador.java

```diff
@@ -0,0 +1,490 @@
+     * flag for simulation step
+     */
+    boolean isSimulationStep = false;
+
+    /**
+     * Sleep after each connection try. Unit: [ms].
+     */
+    private final static long SLEEP_AFTER_ATTEMPT = 1000L;
+
+    /**
+     * Maximum amount of attempts to connect to CARLA simulator.
+     */
+    private int connectionAttempts = 5;
+
+    /**
+     * Carla simulator client port
+     */
+    private int carlaSimulatorClientPort = -1;
+
+    /**
+     * The process for running the connection bridge client
+     */
+    private Process connectionProcess = null;
+
+    /**
+     * Creates a new {@link CarlaAmbassador} object.
+     *
+     * @param ambassadorParameter includes parameters for the CARLA Ambassador.
+     */
+    public CarlaAmbassador(AmbassadorParameter ambassadorParameter) {
+        super(ambassadorParameter);
+        try {
+            // read the CARLA configuration file
+            carlaConfig = new ObjectInstantiation<>(CarlaConfiguration.class, log)
+                    .readFile(ambassadorParameter.configuration);
+        } catch (InstantiationException e) {
+            log.error("Configuration object could not be instantiated: ", e);
+        }
+
+        log.info("carlaConfig.updateInterval: " + carlaConfig.updateInterval);
+
+        // check the carla configuration
+        checkConfiguration();
+    }
+
+    /**
+     * Check the updateInterval is validated.
+     */
+    private void checkConfiguration() {
+        if (carlaConfig.updateInterval <= 0) {
+            throw new RuntimeException("Invalid carla interval, should be >0");
+        }
+    }
+
+    /**
+     * Creates and sets new federate executor.
+     *
+     * @param host name of the host (as specified in /etc/hosts.json)
+     * @param port port number to be used by this federate
+     * @param os   operating system enum
+     * @return FederateExecutor.
+     */
+    @Nonnull
+    @Override
+    public FederateExecutor createFederateExecutor(String host, int port, CLocalHost.OperatingSystem os) {
+        // CARLA needs to start the federate by itself, therefore we need to store the
+        // federate starter locally and use it later
+        federateExecutor = new ExecutableFederateExecutor(descriptor, getCarlaExecutable("CarlaUE4"),
+                getProgramArguments(port));
+        this.carlaSimulatorClientPort = port;
+        return new NopFederateExecutor();
+    }
+
+    /**
+     * Get CARLA simulator executable file location
+     * 
+     * @param executable the name of carla executable file
+     * @return the path to CarlaUE4 executable file
+     */
+    String getCarlaExecutable(String executable) {
+        String carlaHome = null;
+        if (carlaConfig.carlaUE4Path != null) {
+            carlaHome = carlaConfig.carlaUE4Path;
+            log.info("use carla path from configuration file: " + carlaHome);
+        } else if (System.getenv("CARLA_HOME") != null) {
+            carlaHome = System.getenv("CARLA_HOME");
```

Comment:
Is that possible to add log info if the UE4Path is detected from environmental variable?
Also, I am not sure the reason why the first priority of detecting UE4Path is from configure, not environmental variable, could you explain what you concern in this part?

Reply:
The log info has been added when the UE4Path is detected from environment variable.
I initially thought the users can just download the released CARLA and may put the CARLA folder in any place. They just need to specify the CARLA path in the configuration file and do not have to add a new environment variable. 
I changed the path from environmental variable as the first priority.